pkg/rcmanager/node.go

```diff
@@ -0,0 +1,219 @@
+/*
+  Copyright 2021 The Hybridnet Authors.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package rcmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	networkingv1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/constants"
+	"github.com/oecp/rama/pkg/utils"
+	apiv1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	runtimeutil "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+)
+
+const ReconcileNode = "ReconcileNode"
+
+// Full update. Update remote vtep expect status
+func (m *Manager) reconcileNode() error {
+	klog.Infof("[RemoteCluster] Starting reconcile node from cluster %v", m.ClusterName)
+	nodes, err := m.NodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	vteps, err := m.RemoteVtepLister.List(utils.SelectorClusterName(m.ClusterName))
+	if err != nil {
+		return err
+	}
+
+	add, update, remove := m.diffNodeAndVtep(nodes, vteps)
+	var (
+		wg  sync.WaitGroup
+		cur = metav1.Now()
+	)
+	wg.Add(3)
+	go func() {
+		defer wg.Done()
+		for _, v := range add {
+			vtep, err := m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Create(context.TODO(), v, metav1.CreateOptions{})
+			if err != nil {
+				klog.Warningf("Can't create remote vtep in local cluster. err=%v. remote vtep name=%v", err, v.Name)
+				continue
+			}
+			vtep.Status.LastModifyTime = cur
+			_, err = m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().UpdateStatus(context.TODO(), vtep, metav1.UpdateOptions{})
+			if err != nil {
+				runtimeutil.HandleError(err)
+			}
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		for _, v := range update {
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				vtep, err := m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Update(context.TODO(), v, metav1.UpdateOptions{})
```

Comment:
ditto.

Reply:
Now time `reconcileNode` always return nil after `diffNodeAndVtep`. How about using a global err to indicate that there are some error in `add` or `delete` or `update` goroutine and return it so if there are some conflicts, reconcile will happened.

Reply:
I don't think what I said will cause an "error" in your code, but something strange will happen. Maybe both of the operations will success but one of them will seems haven't been executed.

Reply:
Do you mean one of the two reconciling may modify in an old replica because another has applied yet? If so, API server uses optimistic concurrency control to deal with this possible conflicts and thus will return an error from `Update` function. (Anyway, the `RetryOnConflict` is incorrect usage and is useless)

Reply:
I'm not wondering if `Update` function is work correctly, but the remote vtep you used to update. 

If you have get the same remote vtep CR first, then you make some changes to them somehow before both Update operation. In ipinstance reconciling, you just update the `EndpointIPList`, while you update the `vtepIP` and `vtepIP` in node reconciling. 

Now we will find that the two remote vtep CRs you use to Update is different, only the last one being updated will become the result, while no conflict will happened.

Reply:
> I'm not wondering if `Update` function is work correctly, but the remote vtep you used to update.
> 
> If you have get the same remote vtep CR first, then you make some changes to them somehow before both Update operation. In ipinstance reconciling, you just update the `EndpointIPList`, while you update the `vtepIP` and `vtepIP` in node reconciling.
> 
> Now we will find that the two remote vtep CRs you use to Update is different, only the last one being updated will become the result, while no conflict will happened.

This will never happen because there is version checking before `Update` an exist object, if the version is changed, `Update` will not success.